Gambot.Modules.Roll/DiceEvaluator.cs

```diff
@@ -0,0 +1,271 @@
+        }
+
+        private IEnumerable<int> KeepLowest(int numberOfDice, int numberOfSides, int numberToKeep)
+        {
+            IEnumerable<int> rolls = RollDice(numberOfDice, numberOfSides);
+            return rolls.OrderBy(r => r).Take(numberToKeep);
+        }
+
+        private string _Symbols = "khld*/+-()";
+        private IEnumerable<DiceToken> Tokenize(string expression)
+        {
+            List<DiceToken> tokens = new List<DiceToken>();
+            var expr = expression.ToLower();
+
+            var i = 0;
+            while (i < expr.Length)
+            {
+                if (_Symbols.Contains(expr[i]))
+                {
+                    var token = new DiceToken() { IsNumber = false, StringValue = expr[i].ToString() };
+                    if (tokens.Count > 0)
+                    {
+                        token.PrecedingToken = tokens[tokens.Count - 1];
+                        tokens[tokens.Count - 1].ProceedingToken = token;
+                    }
+                    tokens.Add(token);
+                }
+                else if (Char.IsDigit(expr[i]))
+                {
+                    var start = i;
+                    do { ++i; } while (i < expr.Length && Char.IsDigit(expr[i]));
+
+                    var token = new DiceToken() { IsNumber = true, IntValue = Convert.ToInt32(expr.Substring(start, i - start)) };
+                    if (tokens.Count > 0)
+                    {
+                        token.PrecedingToken = tokens[tokens.Count - 1];
+                        tokens[tokens.Count - 1].ProceedingToken = token;
+                    }
+                    tokens.Add(token);
+                    continue;
+                }
+                ++i;
+            }
+
+            var tmp = new List<DiceToken>();
+            foreach (var token in tokens)
+            {
+                if (token.StringValue == "d" && !(token.PrecedingToken != null && token.PrecedingToken.IsNumber))
+                {
+                    tmp.Add(new DiceToken() { IsNumber = true, IntValue = 1 });
+                }
+
+                tmp.Add(token);
+
+                if ((token.StringValue == "h" || token.StringValue == "k" || token.StringValue == "l") && !(token.ProceedingToken != null && token.ProceedingToken.IsNumber))
+                {
+                    tmp.Add(new DiceToken() { IsNumber = true, IntValue = 1 });
+                }
+            }
+            tokens.Clear();
+
+            for (var tmpIdx = 0; tmpIdx < tmp.Count(); ++tmpIdx)
+            {
+                tmp[tmpIdx].ProceedingToken = null;
+                tmp[tmpIdx].PrecedingToken = (tmpIdx == 0) ? null : tmp[tmpIdx - 1];
+
+                if (tmpIdx > 0)
+                {
+                    tmp[tmpIdx - 1].ProceedingToken = tmp[tmpIdx];
+                }
+            }
+
+            return tmp;
+        }
+
+        private bool IsHigherPrecedence(string first, string second)
```

Comment:
This is an odd implementation. You're first slicing out the parens, then doing two `IndexOf`s and comparing them. Having a string of operators in their precedence order isn't the worst idea, but having to `Substring` out the parens comes off as a bit cryptic.